### ISSUE TITLE
HYBRID2: Hybrid Merkle Trees and ZigZag layers

### ISSUE THREAD

**Issue by porcuquine** [filecoin-project/rust-fil-proofs#668]
### Description
This description is resurrected from #531, which was previously scoped down to only implement blake2s in circuits. It looks like this may indeed be worth doing — still pending clear calculations. Briefly, I think my previous calculations were oversimplified, and a rectified übercalc will likely show the hybrid is worthwhile, so I'm reconstituting the original issue.

NOTE: the final solution may be more elaborate than described below. Specifically, we may want the ability to vary the balance per layer. Let's make sure to consider that before beginning work on this, even if the initial implementation is constant across layers.

---
>  Pending calculations verifying a projected solution to the performance equation, implement a hybrid hashing solution for merkle trees. The idea is to allow a tunable public parameter (a tree depth) specifying the portion of the tree which should use blake2s instead of pedersen hashes. This will allow us to trade proof size for speed. This is exactly the lever we need to take advantage of our surplus proof-size budget and apply it to cpu-time performance.

Here's an approach I think can work. If you have a better idea or aren't sure about this, let's discuss before beginning:

  - Implement a new hybrid merkle tree type, say `HybridMerkleTree`, parameterized by two `Hash` types.
  - For this description, call the types `RootHash` and `LeafHash`.
  - Include a field called `root_tree_height` corresponding to the number of rows which should be contained in the `RootHash` tree.
  - The new type should implement all the methods of `merkle_light::MerkleTree` (or at least all the ones we use).
  - When constructing a `HybridMerkleTree`, create `N` `MerkleTree<LeafHash,…>` trees.
  - The root of each of these leaf trees will be one of the leaves of the single `HybridMerkleTree<RootHash,…>` tree.
  - The leaves of all the leaf trees will be the leaves of the hybrid tree.
  - The root of the root tree will be the root of the hybrid tree.

The rest should follow naturally from this description. Example:
 - Say you have a `HybridMerkleTree` of height 10.
 - `root_tree_height = 3`: the root tree will have three rows, and the leaf row will have 4 leaves.
 - Therefore, there will be 4 leaf trees.
 - Each leaf tree will be of height 8.
 - Root tree height + leaf tree height = hybrid tree height - 1.
 - The '- 1` is for the overlap between (unhashed) root-tree leaves and leaf-tree roots, which are the same value.

Once `HybridMerkleTree` is implemented:
 - Refactor the plumbing so `ZigZagDrgPoRep` allows a hybrid tree.  (yes, adding the extra type parameter will be annoying).
 - Add `root_tree_height` to `layered_drgporep::PublicParams` — default 0.
 - This implies that to mimic the behavior of our current `ZigZagDrgPorep`, the `LeafHash` should be `DefaultTreeHasher` (`PedersenHasher`).
 - For consistency, maybe we should also create a type alias for `DefaultRootTreeHasher = Blake2sHasher`.

Last but not even vaguely least, extend circuits to make this work. This probably belongs in a follow-up. Feel free to skip and create a new issue for the circuit component once the vanilla part is done.

Optional for now, but will be needed immediately in a follow-up if not implemented now:
  - Extend zigzag example to allow specification of a root hash and root tree height, using them in the obvious way.

I think that's about it. A lot, I know, but I think it's mostly mechanical if you understand the design.

### Acceptance criteria

  - Everything above is implemented as described (modulo naming, but let's discuss proposed changes).

### Risks + pitfalls
  - There will be a lot of mechanical refactoring in adding the type parameters.
  - I probably failed to anticipate something particularly hairy.
  - I *think* if you stay close to the design, things should shake out — so if it feels like a lot of deviation and innovation beyond dodging the type system is needed, let's discuss.

### Where to begin
 
I'm tempted to change it up and say, 'Sew, a needle pulling thread,' but we should probably keep it simple and start at the beginning, as described above.

**Comment by dignifiedquire** [comment]
we will not be using hybrid merkle trees for now
